### ISSUE TITLE
Consolidate commit message specification

### ISSUE THREAD

**Issue by mih** [datalad/datalad#3896]
https://github.com/datalad/datalad/issues/3316 is asking for `-m/--message` for all commiting commands. There are quite a few, hence this would be a project of some size. https://github.com/datalad/datalad/pull/3883 already started, and https://github.com/datalad/datalad/pull/3893 suggests to add special meaning to a `-m` without an option. There is also: https://github.com/datalad/datalad/issues/3882

It would be good to thing to think this through a bit more and then make one consolidated effort (or at least a plan for such).

**Keep it minimal**: we have plenty of command that need message opts, but `save` as two (`-m` and `-F`). I propose to kill `-F` and replace it with a `file:` argument prefix detection than is then accessible to all commands

**Centralize**: Message file handling should move out of `save` and be extended to a  common helper that deal with any message handling. I think it is worth pondering whether we want a class/function that can be used internally to signal and handle interactive message input, like approached in #3893, possibly with templating for subdatasets as done in https://github.com/datalad/datalad-revolution/pull/71. But there is another templating aspect too: https://github.com/datalad/datalad/issues/2100

**What about tagging?**: https://github.com/datalad/datalad/issues/3371

**Comment by yarikoptic** [comment]
Not sure about killing -F. On one hand indeed it is less if passing things around, but in the other - obfuscation. I wonder if we could resolve it programmatically. So both are present, but -F just acts via callback to the option specification itself and populates `message` option, (may be even somehow that so code itself doesn't need to even list it in the signature).

**Comment by mih** [comment]
Hmm, you did not convince me that the implied obfuscation is more of a problem than the implied complexity of introducing two options to a large number of commands (instead of just one).

If you don't like `file:` prefix, we can make it a proper `file://` URL, which would impose that any message files need to be given via their absolute path. I found that annoying, but if that mitigates your concerns, so be it.

**Comment by bpoldrack** [comment]
First-off, since this is labeled a question, I agree that message (file-) handling should be centralized and minimal in terms of API pollution, if we want to be able to specify messages for more commands than just `save`.

This would need a - possibly interactive - direct message spec as well as predefined messages from a file, templates, tags and a mechanism to deal with subdatasets in case of multiple commits.

Re `-F`:
I don't quite get the obfuscation aspect. What gets obfuscated?
To me it makes a lot of sense to specify either a file/template or the message itself via the same parameter and let the value make the difference. If there was a case where one could specify both simultaneously, that'd be a different story. But `-m` and `-F` are mutually exclusive anyway ATM.

However, I think it pretty much depends on how to specify templates and deal with subdatasets. We might end up with a scheme where `-m` would specify either a message or a "mode" and `-F` the possibly required file for that mode (template or just message or some config to write to a subdataset?). Also, one might want to have a similar mechanism for tags. So, overall we might actually need a separate file parameter. I'm not quite there yet, but I'd decide about the faith of a particular parameter, when we have a concept on how to tackle those other aspects generally.

**Comment by mih** [comment]
Just to comment on the API complexity aspect: I have grown fond of using "mode" argument values over multiple dedicated options that somehow always seem to turn in combinatorial issues. In this case, I see no issues folding the mode and the value into `-m`:

- `"just a message"`
- `file:.git/datalad/prev-message`
- `stdin:`
- `template:/how/we/do/commit/msgs/here.txt` for templating, as done in https://github.com/datalad/datalad-revolution/pull/71

**Comment by bpoldrack** [comment]
Yes, absolutely agree in general. What I'm not clear about is the cases where we commit to several levels of a dataset hierarchy and how to specify those cases.  Mostly because I'm not sure what we actually want allow for. So, IF that is leading to the need of an additional option anyway, THEN it would be the consistent approach to keep using that option in general. If not, then I completely agree with that prefix approach.

Side-note: If we were to actually use `file://` we can even have arbitrary URLs, which can become convenient (particularly in case of templates) for institutional templates, that are then just hosted somewhere.

This in turn leads to just allow for URLs where you, @mih, used paths. (so - "direct files" as well as templates). In that case `file:` and `file://` wouldn't actually be alternatives. Plus: `file:` and `template:` can possibly be merged and auto-detected by the presence of whatever markers to be used in a template. However, no strong opinion on that merge - might be clearer to separate them.

So, maybe:

- `"just a message"`
- `predef: file://some/where.txt`
- `template: https://example.com/fancy_template.txt`
- `stdin:`

**Comment by bpoldrack** [comment]
Apart from the question how exactly to specify those, I guess we want some configs to be queried in case of no passed message. This could be command-based, like `datalad.install.message="template: ...."`? Not quite sure yet, whether the command should be the anchor point, but sth. similar could be the way to have a general setup as well as the way to tell subdatasets what to use.

**Comment by mih** [comment]
@yarikoptic What do you think?

**Comment by mih** [comment]
We managed to agree on something here (in a call), and not write the conclusion down. I think it was "have a single `--message`, and <forgotten>". :disappointed:

**Comment by mih** [comment]
Two years went by... Is the plan with just having `--message` a GO now?
